lib/pbench/server/api/resources/query_apis/search_api.py

```diff
@@ -0,0 +1,201 @@
+                'run.controller': 'dhcp31-171.perf.lab.eng.bos.redhat.com',
+                'run.name': 'pbench-user-benchmark_npalaska-dhcp31-171_2021.07.12T14.08.46',
+                'host_tools_info.hostname.test1': 'dhcp31-171',
+                'host_tools_info.hostname.test2': 'dhcp31-195'
+            }
+        """
+        if type(es_json) is dict:
+            for key in es_json:
+                while prefix[-2:-1] == ".":
+                    prefix = prefix[:-1]
+                self.flatten_json(es_json[key], prefix + key + ".", out)
+        elif type(es_json) is list:
+            i = 0
+            for element in es_json:
+                if type(element) is dict:
+                    self.flatten_json(element, prefix + ".", out)
+                else:
+                    self.flatten_json(element, prefix + str(i) + ".", out)
+                i += 1
+        else:
+            out[prefix[:-1]] = es_json
+        return out
+
+    def assemble(self, json_data: Dict[AnyStr, Any]) -> Dict[AnyStr, Any]:
+        """
+        Construct a pbench search query based on a pattern matching given "search_term" parameter
+        within a specified date range and which are either owned by a specified username,
+        or have been made publicly accessible.
+
+        {
+            "user": "username",
+            "start": "start-time",
+            "end": "end-time"
+            "search_term": "string_pattern_to_search"
+            "fields": list(interested run document fields returned from IndexMapping API)
+        }
+
+        json_data: JSON dictionary of type-normalized parameters
+            user: specifies the owner of the data to be searched; it need not
+                necessarily be the user represented by the session token
+                header, assuming the session user is authorized to view "user"s
+                data. If "user": None is specified, then only public datasets
+                will be returned.
+
+            "start" and "end" are datetime objects representing a set of Elasticsearch
+                run document indices in which to search.
+
+            "search_term": string pattern to match against all the datasets owned by a specified
+                    username within a given time range
+
+            "fields": List of run document fields end user is interested in. E.g.
+                    ["@timestamp", "run.controller", "run.name", ...]
+        """
+        user = json_data.get("user")
+        start = json_data.get("start")
+        end = json_data.get("end")
+        search_term = json_data.get("search_term")
+        # If the search_term paramter is not specified we will return all the elastisearch hits
+        if search_term is None:
+            search_term = ""
+        selected_fields = json_data.get("fields")
+        # If no fileds are specified, the query will return all the fields from elasticsearch hits
+        if selected_fields is None:
+            selected_fields = []
```

Comment:
Is that the way it works? `"_source": {"include": []}` returns *everything* instead of *nothing*?? (OK, yes, that is the way it works. Although, interestingly, specifying a field name that doesn't exist *will* return `"_source": {}`.)

Reply:
Which is an expected output right,? If we specify non existent fields with some existent fields its only gonna return `_source` with those existent fields ignoring the non-existent ones

Reply:
If you specify a (non-empty) list of stuff to return, then ES will return only the stuff in that list, and only if it is present in the data.  Thus, if you specify only a non-existent field, then you get no data back.  (None of this is surprising.)

The interesting facet arises when you _specify_ a list which **empty**.  Apparently, ES is treating that the same as _not_ specifying anything, and that is slightly surprising.  _Specifying_ that nothing be returned seems like it should result in nothing being returned, but this is an unusual request, and there are other ways of making it (like asking for zero results), so I guess the ES folks decided it would be better or more useful to treat it the other way.

Reply:
Yeah for ES be like, my job is to return the data how can someone specify zero data to return, I am returning everything :)